io.openems.edge.levl.controller/src/io/openems/edge/levl/controller/LevlControlRequest.java

```diff
@@ -0,0 +1,140 @@
+package io.openems.edge.levl.controller;
+
+import com.google.gson.JsonObject;
+import io.openems.common.exceptions.OpenemsError;
+import io.openems.common.exceptions.OpenemsError.OpenemsNamedException;
+import io.openems.common.jsonrpc.base.JsonrpcRequest;
+import io.openems.common.utils.JsonUtils;
+
+import java.time.Clock;
+import java.time.LocalDateTime;
+import java.util.Objects;
+
+public class LevlControlRequest {
+	public static final String METHOD = "sendLevlControlRequest";
+	public static final int QUARTER_HOUR_SECONDS = 900;
+	protected static Clock clock = Clock.systemDefaultZone();
+	protected int sellToGridLimitW;
+	protected int buyFromGridLimitW;
+	protected String levlRequestId;
+	protected String timestamp;
+	protected long energyWs;
+	protected LocalDateTime start;
+	protected LocalDateTime deadline;
+	protected int levlSocWh;
+	protected double socLowerBoundPercent;
+	protected double socUpperBoundPercent;
+	protected double efficiencyPercent;
+	protected boolean influenceSellToGrid;
+	
+	public LevlControlRequest() {
+		
+	}
+	
+	public LevlControlRequest(JsonObject params) throws OpenemsError.OpenemsNamedException {
+		try {
+			this.parseFields(params);
+		} catch (NullPointerException e) {
+			throw OpenemsError.JSONRPC_INVALID_MESSAGE.exception("missing fields in request: " + e.getMessage());
+		} catch (NumberFormatException e) {
+			throw OpenemsError.JSONRPC_INVALID_MESSAGE.exception("wrong field type in request: " + e.getMessage());
+		}
+		if (this.efficiencyPercent < 0) {
+			throw OpenemsError.JSONRPC_INVALID_MESSAGE.exception("efficiencyPercent must be > 0");
+		}
+		if (this.efficiencyPercent > 100) {
+			throw OpenemsError.JSONRPC_INVALID_MESSAGE.exception("efficiencyPercent must be <= 100");
+		}
+	}
+
+	//Just for testing
+	public LevlControlRequest(int sellToGridLimitW, int buyFromGridLimitW, String levlRequestId, String timestamp,
```

Comment:
package level visibility should be good enough then, isn't it?

Reply:
`protected` is more than package level visibility, see https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html

But `protected` is good as well :)